### PR TITLE
Dropping support for python 3.7 

### PR DESCRIPTION
**Pull request by bsipocz**
EDIT: This had to be broke up into 2 separate PRs to be able to add two changelog entries. See #11935 for dropping numpy

This is in agreement with APE18 and the email thread: https://groups.google.com/g/astropy-dev/c/Z9TPHm8DSmo/m/TgGT6BwFBAAJ?pli=1

Note: the changelog entry is not right, we will need two separate lines, both having the PR number at the end of the line when rendered.

I also expect some of the CI builds to fail, as I may not get the image names right at the first try.

- [ ] Restart CI after Debian updates upstream w.r.t. wcslib 7.6. Should happen by 2021-09-01.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by nstarman** [comment]
Wooooh! Walrus operator!

```
               ___
            .-9 9 `\
          =(:(::)=  ;
            ||||     \
            ||||      `-.
           ,\|\|         `,
          /                \
         ;                  `'---.,
         |                         `\
         ;                     /     |
         \                    |      /
  jgs     )           \  __,.--\    /
       .-' \,..._\     \`   .-'  .-'
      `-=``      `:    |   /-/-/`
                   `.__/
```

**Comment by pllim** [comment]
Re: walrus operator -- Apparently not everyone was excited about it. See #9606

**Comment by pllim** [comment]
@bsipocz , thanks! I'll review once we decide whether to merge this (back) with #11935 or not.

**Comment by mhvk** [comment]
The walrus will just come in slowly, I suspect, as people make commits that use it. Not quite like f-strings, where it is (relatively) easy to search for relevant places (and which were a really nice addition indeed!).

File: .circleci/config.yml
Comment (bsipocz) [inline .circleci/config.yml:9]: maybe switch directly to py39 instead of 38?

**Comment by mhvk** [comment]
Just had a look through the changes, and they all look good to me - indeed, rather little is affected. To me, it would seem to make sense to have the matplotlib images be built on 3.9 rather than 3.8, to avoid churn when the minimum version becomes 3.9 and since we test all dependencies with 3.8 elsewhere anyway.

File: .circleci/config.yml
Comment (pllim) [reply .circleci/config.yml:9]: @astrofrog pointed to https://docs.astropy.org/en/latest/development/testguide.html#running-image-tests but someone with server access still needs to generate and upload the new image(s).

**Comment by olebole** [comment]
@bsipocz The newest wcslib, 7.7 is now uploaded to Debian (already migrated to testing today). astropy 4.3.1 seems happy with it :eyeglasses:

**Comment by pllim** [comment]
Thanks for the update! I restarted the affected jobs... 🤞

**Comment by pllim** [comment]
Did I restart the job too early? I still see this:

`Get:153 .../debian bullseye/main arm64 wcslib-dev arm64 7.4+ds-2 [367 kB]`

**Comment by olebole** [comment]
Bullseye comes with 7.4, and will remain so. You need to switch to testing (or bookworm) instead.

**Comment by pllim** [comment]
@olebole , what we can use is limited by what is provided at https://github.com/uraimo/run-on-arch-action#supported-platforms . I don't see "testing" or "bookworm". Any advise?

**Comment by olebole** [comment]
Hmm, I could backport the new wcslibs to bullseye. This would require that these platforms include the bullseye-backports repository, and one needs to specify the version in apt.

**Comment by pllim** [comment]
@olebole , that might work if we can reapply a patch like this here:

0ff2f4f4bb2493195df42a674f349a8e8fa16edf

**Comment by olebole** [comment]
Yep. that probably would work (with `s/buster/bullseye/`). However, I need a few more days to get that ready, and it needs the formal acceptance by the ftp-masters.

**Comment by pllim** [comment]
@olebole , any update on the status of backport? Thanks!

**Comment by olebole** [comment]
@pllim Thank you for the reminder. [I uploaded it](https://ftp-master.debian.org/new/wcslib_7.7+ds-1~bpo11+1.html), and it now needs approval by our ftp-masters.

**Comment by pllim** [comment]
@olebole , any timeline on when your patch would be accepted and deployed? Thanks!

I am thinking of just merging this PR by the end of today and let the cron job fail until Debian can provide this backport. This is because this PR has been open long enough and I don't want it to have merge conflicts.

**Comment by olebole** [comment]
That is hard to predict. However, as their backports queue currently contains entries that are only one day old, I guess it will happen in the next days. Sorry, but the ftp-master require some patience, even from us Debian Developers :-)

**Comment by pllim** [comment]
Okay, let's just merge then. I'll open a follow-up issue. Thanks for the update! 😄

**Review by pllim — APPROVED** [review]

**Pull request merged**
